File: src/hyperLiquid_ts_deprecated/repositories/dynamo.repository.ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, PutCommand, ScanCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { OrderRecord, NewOrder } from '../models/order';
import * as crypto from "crypto";

const client = new DynamoDBClient({ region: process.env.AWS_REGION || "us-east-1" });
const docClient = DynamoDBDocumentClient.from(client);

const tableName = process.env.DYNAMODB_TABLE_NAME || "hl-orders";

export async function insertOrder(order: NewOrder): Promise<OrderRecord> {
    const id = crypto.randomUUID();
    const createdAt = new Date().toISOString();
    
    const item = {
        id,
        userAddress: order.user_address,
        symbol: order.symbol,
        strategy: order.strategy,
        quantity: order.quantity.toString(),
        orderType: order.order_type,
        price: order.price.toString(),
        oid: order.oid,
        stopLossOid: order.stopLossOid || '',
        stopLossPrice: order.stopLossPrice ? order.stopLossPrice.toString() : '',
        pnl: '',
        status: order.status,
        createdAt: createdAt,
        updatedAt: createdAt,
    };

    await docClient.send(new PutCommand({
        TableName: tableName,
        Item: item
    }));
    
    return entityToOrder(item);
}

function entityToOrder(entity: any): OrderRecord {
    return {
        partitionKey: `${entity.symbol}_${entity.strategy}`,
        id: entity.id,
        user_address: entity.userAddress,
        symbol: entity.symbol,
        strategy: entity.strategy,
        quantity: parseFloat(entity.quantity),
        order_type: entity.orderType,
        price: parseFloat(entity.price),
        oid: entity.oid,
        stopLossOid: entity.stopLossOid || undefined,
        stopLossPrice: entity.stopLossPrice ? parseFloat(entity.stopLossPrice) : undefined,
        pnl: entity.pnl ? parseFloat(entity.pnl) : undefined,
        status: entity.status,
        created_at: new Date(entity.createdAt),
        updated_at: new Date(entity.updatedAt),
    };
}

export async function getOrder(options: {
    symbol?: string;
    strategy?: string;
    id?: string;
    status?: string;
}): Promise<OrderRecord | null> {
    
    let filterExpression = [];
    let expressionAttributeValues: any = {};
    let expressionAttributeNames: any = {};
    
    if (options.id) {
        filterExpression.push("id = :id");
        expressionAttributeValues[":id"] = options.id;
    }
    if (options.symbol) {
        filterExpression.push("#sym = :sym");
        expressionAttributeNames["#sym"] = "symbol";
        expressionAttributeValues[":sym"] = options.symbol;
    }
    if (options.strategy) {
        filterExpression.push("strategy = :strategy");
        expressionAttributeValues[":strategy"] = options.strategy;
    }
    if (options.status) {
        filterExpression.push("#st = :st");
        expressionAttributeNames["#st"] = "status";
        expressionAttributeValues[":st"] = options.status;
    }

    const scanParams: any = {
        TableName: tableName
    };

    if (filterExpression.length > 0) {
        scanParams.FilterExpression = filterExpression.join(" AND ");
        scanParams.ExpressionAttributeValues = expressionAttributeValues;
        if (Object.keys(expressionAttributeNames).length > 0) {
            scanParams.ExpressionAttributeNames = expressionAttributeNames;
        }
    }

    const { Items } = await docClient.send(new ScanCommand(scanParams));
    if (Items && Items.length > 0) {
        return entityToOrder(Items[0]);
    }
    return null;
}

export async function listAllOpenOrders(): Promise<OrderRecord[]> {
    const { Items } = await docClient.send(new ScanCommand({
        TableName: tableName,
        FilterExpression: "#st = :status",
        ExpressionAttributeNames: { "#st": "status" },
        ExpressionAttributeValues: { ":status": "open" }
    }));
    
    return (Items || []).map(entityToOrder);
}

export async function updateOrder(
    id: string,
    updates: Partial<{
        status: string;
        pnl: number;
        oid: string;
        stopLossOid: string;
        stopLossPrice: number;
    }>
): Promise<void> {
    const updateExpressions: string[] = ["updatedAt = :updatedAt"];
    const expressionAttributeValues: any = {
        ":updatedAt": new Date().toISOString()
    };
    const expressionAttributeNames: any = {};

    if (updates.status !== undefined) {
        updateExpressions.push("#st = :status");
        expressionAttributeNames["#st"] = "status";
        expressionAttributeValues[":status"] = updates.status;
    }
    if (updates.pnl !== undefined) {
        updateExpressions.push("pnl = :pnl");
        expressionAttributeValues[":pnl"] = updates.pnl.toString();
    }
    if (updates.oid !== undefined) {
        updateExpressions.push("oid = :oid");
        expressionAttributeValues[":oid"] = updates.oid.toString();
    }
    if (updates.stopLossOid !== undefined) {
        updateExpressions.push("stopLossOid = :stopLossOid");
        expressionAttributeValues[":stopLossOid"] = updates.stopLossOid.toString();
    }
    if (updates.stopLossPrice !== undefined) {
        updateExpressions.push("stopLossPrice = :stopLossPrice");
        expressionAttributeValues[":stopLossPrice"] = updates.stopLossPrice.toString();
    }

    const updateParams: any = {
        TableName: tableName,
        Key: { id },
        UpdateExpression: `SET ${updateExpressions.join(", ")}`,
        ExpressionAttributeValues: expressionAttributeValues,
    };

    if (Object.keys(expressionAttributeNames).length > 0) {
        updateParams.ExpressionAttributeNames = expressionAttributeNames;
    }

    await docClient.send(new UpdateCommand(updateParams));
}

File: src/hyperliquid_python/repositories/dynamo_repository.py
import os
import uuid
import boto3
from datetime import datetime, timezone
from typing import Optional, List, Dict, Any
from models.order import OrderRecord, NewOrder

dynamodb = boto3.resource('dynamodb', region_name=os.environ.get("AWS_REGION", "us-east-1"))
table_name = os.environ.get("DYNAMODB_TABLE_NAME", "hl-orders")
table = dynamodb.Table(table_name)

def entity_to_order(entity: dict) -> OrderRecord:
    return OrderRecord(
        partitionKey=f"{entity.get('symbol')}_{entity.get('strategy')}",
        id=entity.get('id'),
        user_address=entity.get('userAddress'),
        symbol=entity.get('symbol'),
        strategy=entity.get('strategy'),
        quantity=float(entity.get('quantity')),
        order_type=entity.get('orderType'),
        price=float(entity.get('price')),
        oid=entity.get('oid'),
        stopLossOid=entity.get('stopLossOid') if entity.get('stopLossOid') else None,
        stopLossPrice=float(entity.get('stopLossPrice')) if entity.get('stopLossPrice') else None,
        pnl=float(entity.get('pnl')) if entity.get('pnl') else None,
        status=entity.get('status'),
        created_at=datetime.fromisoformat(entity.get('createdAt').replace('Z', '+00:00')),
        updated_at=datetime.fromisoformat(entity.get('updatedAt').replace('Z', '+00:00'))
    )

def insert_order(order: NewOrder) -> OrderRecord:
    order_id = str(uuid.uuid4())
    created_at = datetime.now(timezone.utc).isoformat()
    
    item = {
        'id': order_id,
        'userAddress': order.user_address,
        'symbol': order.symbol,
        'strategy': order.strategy,
        'quantity': str(order.quantity),
        'orderType': order.order_type,
        'price': str(order.price),
        'oid': order.oid,
        'stopLossOid': order.stopLossOid or '',
        'stopLossPrice': str(order.stopLossPrice) if order.stopLossPrice else '',
        'pnl': '',
        'status': order.status,
        'createdAt': created_at,
        'updatedAt': created_at,
    }

    table.put_item(Item=item)
    return entity_to_order(item)

def get_order(
    symbol: Optional[str] = None,
    strategy: Optional[str] = None,
    id: Optional[str] = None,
    status: Optional[str] = None
) -> Optional[OrderRecord]:
    
    filter_expressions = []
    expression_attribute_values = {}
    expression_attribute_names = {}
    
    if id:
        filter_expressions.append("id = :id")
        expression_attribute_values[":id"] = id
    if symbol:
        filter_expressions.append("#sym = :sym")
        expression_attribute_names["#sym"] = "symbol"
        expression_attribute_values[":sym"] = symbol
    if strategy:
        filter_expressions.append("strategy = :strategy")
        expression_attribute_values[":strategy"] = strategy
    if status:
        filter_expressions.append("#st = :st")
        expression_attribute_names["#st"] = "status"
        expression_attribute_values[":st"] = status

    scan_params = {}
    
    if filter_expressions:
        scan_params['FilterExpression'] = " AND ".join(filter_expressions)
        scan_params['ExpressionAttributeValues'] = expression_attribute_values
        if expression_attribute_names:
            scan_params['ExpressionAttributeNames'] = expression_attribute_names

    response = table.scan(**scan_params)
    items = response.get('Items', [])
    
    if items:
        # returns the first one just like JS code
        return entity_to_order(items[0])
    return None

def list_all_open_orders() -> List[OrderRecord]:
    response = table.scan(
        FilterExpression="#st = :status",
        ExpressionAttributeNames={"#st": "status"},
        ExpressionAttributeValues={":status": "open"}
    )
    items = response.get('Items', [])
    return [entity_to_order(item) for item in items]

def update_order(
    id: str,
    status: Optional[str] = None,
    pnl: Optional[float] = None,
    oid: Optional[str] = None,
    stopLossOid: Optional[str] = None,
    stopLossPrice: Optional[float] = None
) -> None:
    update_expressions = ["updatedAt = :updatedAt"]
    expression_attribute_values = {
        ":updatedAt": datetime.now(timezone.utc).isoformat()
    }
    expression_attribute_names = {}

    if status is not None:
        update_expressions.append("#st = :status")
        expression_attribute_names["#st"] = "status"
        expression_attribute_values[":status"] = status
    if pnl is not None:
        update_expressions.append("pnl = :pnl")
        expression_attribute_values[":pnl"] = str(pnl)
    if oid is not None:
        update_expressions.append("oid = :oid")
        expression_attribute_values[":oid"] = str(oid)
    if stopLossOid is not None:
        update_expressions.append("stopLossOid = :stopLossOid")
        expression_attribute_values[":stopLossOid"] = str(stopLossOid)
    if stopLossPrice is not None:
        update_expressions.append("stopLossPrice = :stopLossPrice")
        expression_attribute_values[":stopLossPrice"] = str(stopLossPrice)

    update_params = {
        'Key': {'id': id},
        'UpdateExpression': f"SET {', '.join(update_expressions)}",
        'ExpressionAttributeValues': expression_attribute_values
    }

    if expression_attribute_names:
        update_params['ExpressionAttributeNames'] = expression_attribute_names

    table.update_item(**update_params)
